src/components/exercise/part.cjsx

```diff
@@ -0,0 +1,130 @@
+    onStepCompleted: React.PropTypes.func.isRequired
+    onNextStep: React.PropTypes.func.isRequired
+    getCurrentPanel: React.PropTypes.func.isRequired
+    step: step
+
+    setFreeResponseAnswer: React.PropTypes.func.isRequired
+    setAnswerId: React.PropTypes.func.isRequired
+
+    getReadingForStep: React.PropTypes.func
+    refreshStep: React.PropTypes.func
+    recoverFor: React.PropTypes.func
+
+    review: React.PropTypes.string
+    focus: React.PropTypes.bool
+    courseId: React.PropTypes.string
+    canTryAnother: React.PropTypes.bool
+    canReview: React.PropTypes.bool
+    disabled: React.PropTypes.bool
+
+  getInitialState: ->
+    {id} = @props
+
+    currentPanel: @props.getCurrentPanel(id)
+
+  componentWillMount: ->
+    {id} = @props
+    @updateCurrentPanel(@props) unless @state.currentPanel
+
+  componentWillReceiveProps: (nextProps) ->
+    @updateCurrentPanel(nextProps)
+
+  updateCurrentPanel: (props) ->
+    {id, getCurrentPanel} = props or @props
+    currentPanel = getCurrentPanel(id)
+    @setState({currentPanel}) if currentPanel? and @state.currentPanel isnt currentPanel
+
+  getDefaultProps: ->
+    focus: true
+    review: ''
+    pinned: true
+    canTryAnother: false
+    canReview: false
+
+  refreshMemory: ->
+    {id, taskId} = @props
+
+    {index} = @props.getReadingForStep(id, taskId)
+    @props.refreshStep(index, id)
+
+  tryAnother: ->
+    {id} = @props
+    @props.recoverFor(id)
+
+  onFreeResponseContinue: (state) ->
+    {id} = @props
+    {freeResponse} = state
+    @props.setFreeResponseAnswer(id, freeResponse)
+
+  onMultipleChoiceAnswerChanged: (answer) ->
+    {id} = @props
+    @props.setAnswerId(id, answer.id)
+
+  getReviewProps: ->
+    reviewProps = _.omit(@props, NOT_REVIEW_PROPS)
+    reviewProps.onContinue = @props.onNextStep
+    reviewProps.refreshMemory = @refreshMemory
+    reviewProps.tryAnother = @tryAnother
+
+    reviewProps
+
+  getMultipleChoiceProps: ->
+    multipleChoiceProps = _.omit(@props, NOT_MULTIPLE_CHOICE_PROPS)
+    multipleChoiceProps.onAnswerChanged = @onMultipleChoiceAnswerChanged
+
+    multipleChoiceProps
+
+  getFreeResponseProps: ->
+    freeResponseProps = _.omit(@props, NOT_FREE_RESPONSE_PROPS)
+    freeResponseProps.onContinue = @onFreeResponseContinue
+
+    freeResponseProps
+
+  getTeacherReadOnlyProps: ->
+    teacherReadOnlyProps = _.omit(@props, NOT_TEACHER_READ_ONLY_PROPS)
+    teacherReadOnlyProps.onContinue = @props.onNextStep
+    teacherReadOnlyProps.controlButtons = false
+    teacherReadOnlyProps.type = 'teacher-review'
+
+    teacherReadOnlyProps
+
+  # add get props methods for different panel types as needed here
+
+  render: ->
+    {id, step, getWaitingText, waitingText} = @props
+    {currentPanel} = @state
+
+    # panel is one of ['review', 'multiple-choice', 'free-response', 'teacher-read-only']
+    getPropsForPanel = camelCase "get-#{currentPanel}-props"
+    cardProps = @[getPropsForPanel]?()
+
```

Comment:
This makes me nervous because if a bad `getCurrentPanel` prop is supplied, it's possible the `currentPanel` might refer to a non-existent function and cause a JS error that would stop rendering.  

Maybe we could use `cardProps =  _.result(@, "get-#{currentPanel}-props")`  That would tolerate non-existent properties and simply return null rather than causing an exception.